### PR TITLE
NetworkitBinaryGraph is now able to save Edge Indices

### PR DESCRIPTION
**Pull request by Relux-the-Relux**
Addressing #495 .

### CODE REVIEW

**Comment by avdgrinten** [comment]
This is a nice feature.

We also need to update the documentation of the file format to describe how the edge IDs are stored, in:
https://github.com/networkit/networkit/blob/master/networkit/cpp/io/NetworkitBinaryGraph.md

Do we need to bump the version? Can't the version 2 reader read a graph with edge IDs, too? As far as I can see, the old reader will just ignore the edge IDs but it will continue to work, right?

**Comment by Relux-the-Relux** [comment]
> Do we need to bump the version? Can't the version 2 reader read a graph with edge IDs, too? As far as I can see, the old reader will just ignore the edge IDs but it will continue to work, right?

No, I didn't bump the version it was just  error in the documentation(from my side) I fixed it it is still nkbg02

**Comment by Relux-the-Relux** [comment]
Also going to update the documentation

**Comment by avdgrinten** [comment]
This does not solve the pickle issue yet -- we still need to implement logic to pickle and unpickle graphs in Python.

**Comment by avdgrinten** [comment]
Your implementation stores the graph in a file and then saves the file name to implement the pickle operation. That is not really what users expect -- the pickle operation should store the graph itself. One option would be reading the file back into a Python 3 `bytes` object.

**Comment by Relux-the-Relux** [comment]
Ok made the change that it now reads from the file into memory. However, this causes 4 hard disk accesses to get and then set the states (2 for getstate and 2 for setstate).

maybe we should implement a version of BinaryReader/Writer that take/output binary strings? Unsure how often get/setstate will get called so that might be a non-issue.

File: include/networkit/io/NetworkitBinaryReader.hpp [hunk 1]
```diff
@@ -24,12 +27,17 @@ class NetworkitBinaryReader final : public GraphReader {
     NetworkitBinaryReader(){};
 
     Graph read(const std::string &path) override;
+    Graph readState(const std::string &data);
```
Comment (avdgrinten) [inline include/networkit/io/NetworkitBinaryReader.hpp:30]: Rename to `readFromBuffer`. Let this take a `std::vector<uint8_t>` instead of a `std::string`.

File: include/networkit/io/NetworkitBinaryWriter.hpp [hunk 2]
```diff
@@ -30,13 +30,19 @@ class NetworkitBinaryWriter final : public GraphWriter {
 
 public:
     NetworkitBinaryWriter(uint64_t chunks = 32,
-                          NetworkitBinaryWeights weightsType = NetworkitBinaryWeights::autoDetect);
+                          NetworkitBinaryWeights weightsType = NetworkitBinaryWeights::autoDetect,
+                          bool preserveEdgeIndex = false);
 
     void write(const Graph &G, const std::string &path) override;
+    std::string writeState(const Graph &G);
```
Comment (avdgrinten) [inline include/networkit/io/NetworkitBinaryWriter.hpp:37]: Rename to `writeToBuffer` and let this return a `std::vector<uint8_t>`.

File: networkit/cpp/io/NetworkitBinaryGraph.md [hunk 3]
```diff
@@ -20,25 +20,33 @@ struct Header {
...
-        - 1 = Weights are signed integers
-        - 2 = Weights are doubles
-        - 3 = Weights are floats
+    - Bit 1-3 : the weight format of the graph. Below are the possible formats:
+		- 0 = Graph is unweighted
+        - 1 = Weights are unsigned integers
+        - 2 = Weights are signed integers
+        - 3 = Weights are doubles
+        - 4 = Weights are floats
+	- Bit 4 = proper edge indexing
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryGraph.md:38]: "proper edge indexing" -> "edge indices are present". Remove the unrelated changes above this one.

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 4]
```diff
@@ -7,23 +7,44 @@
...
+
+template <>
+const char *getIterator(const NetworKit::MemoryMappedFile &source) {
+    return source.cbegin();
+}
+
+template <>
+const char *getIterator(const std::string &source) {
+    return source.data();
+}
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:28]: Do not use specialization here. Use multiple function overloads instead (basically, just drop the `template <>`).

Comment (avdgrinten) [reply networkit/cpp/io/NetworkitBinaryReader.cpp:28]: These functions are local to the file and should be in an anonymous namespace.

Comment (avdgrinten) [reply networkit/cpp/io/NetworkitBinaryReader.cpp:28]: Since these functions return a `const char *`, `accessData()` would be a better name.

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 5]
```diff
@@ -89,12 +116,15 @@ Graph NetworkitBinaryReader::read(const std::string& path) {
...
-    const char *adjIt = mmfile.cbegin() + header.offsetAdjLists;
-    const char *transpIt = mmfile.cbegin() + header.offsetAdjTranspose;
-    const char *adjWghtIt = mmfile.cbegin() + header.offsetWeightLists;
-    const char *transpWghtIt = mmfile.cbegin() + header.offsetWeightTranspose;
+    const char *adjIt = getIterator(source) + header.offsetAdjLists;
+    const char *transpIt = getIterator(source) + header.offsetAdjTranspose;
+    const char *adjWghtIt = getIterator(source) + header.offsetWeightLists;
+    const char *transpWghtIt = getIterator(source) + header.offsetWeightTranspose;
+    const char *adjIdIt = getIterator(source) + header.offsetAdjIdLists;
+    const char *transpIdIt = getIterator(source) + header.offsetAdjIdTranspose;
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:125]: Save the result of `getIterator(source)` to a local variable to avoid duplication here.

File: networkit/cpp/io/NetworkitBinaryWriter.cpp [hunk 6]
```diff
@@ -249,74 +282,97 @@ void NetworkitBinaryWriter::write(const Graph &G, const std::string &path) {
...
     }
+
+    if(preserveEdgeIndex && weightFormat != nkbg::WEIGHT_FORMAT::NONE) {
+        header.offsetAdjIdLists = header.offsetWeightTranspose
+            + (chunks - 1) * sizeof(uint64_t)//transpWeightOffsets
+            + transpWghtOffsets.back();
+        header.offsetAdjIdTranspose = header.offsetAdjIdLists
+            + (chunks - 1) * sizeof(uint64_t)
+            + adjIndexOffsets.back();
+    } else if(preserveEdgeIndex) { //Even if there is no weight data, offset for weights are still written
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryWriter.cpp:293]: `//Even if there is no weight data, offset for weights are still written` -> Why is this done?

**Review by avdgrinten — CHANGES_REQUESTED** [review]

File: networkit/cpp/io/NetworkitBinaryWriter.cpp [hunk 7]
```diff
@@ -17,12 +18,23 @@
...
+}
+
+std::string NetworkitBinaryWriter::writeState(const Graph &G) {
+    std::stringstream outdata;
+    writeData(outdata, G);
+    return outdata.str();
+}
+
+template <class T>
+void NetworkitBinaryWriter::writeData(T &outStream, const Graph &G){
```
[the remark this answers is not in the document]
Comment (Relux-the-Relux) [reply networkit/cpp/io/NetworkitBinaryWriter.cpp:37]: In this case, the MemoryMappedFile functionality of Networkit would need to be expanded as it currently can only put existing files into memory (and will fail when opening empty files) and has currently no write functionality.

File: networkit/cpp/io/NetworkitBinaryWriter.cpp [hunk 8]
```diff
@@ -249,74 +282,97 @@ void NetworkitBinaryWriter::write(const Graph &G, const std::string &path) {
         header.offsetWeightLists = 0;
         header.offsetWeightTranspose = 0;
     }
+
+    if(preserveEdgeIndex && weightFormat != nkbg::WEIGHT_FORMAT::NONE) {
```
[the remark this answers is not in the document]
Comment (avdgrinten) [reply networkit/cpp/io/NetworkitBinaryWriter.cpp:286]: Maybe we should just have a `nextOffset` variable that starts at 0 (or at the size of the headers) and that we increment whenever we "allocate" space for more data within the file? That approach would be less complicated than calculating later offsets based on earlier offsets.

File: networkit/cpp/io/NetworkitBinaryWriter.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Relux-the-Relux) [reply networkit/cpp/io/NetworkitBinaryWriter.cpp:37]: yes, it works since string to std::vector<uint8_t> is a easy conversion. So no problem on the c++ side. I currently just need to test with cython since the documentation recommends using std::string when interfacing bytes data with python.

File: networkit/cpp/io/NetworkitBinaryWriter.cpp [hunk 9]
```diff
@@ -17,12 +18,24 @@
...
+    writeData(outfile, G);
+    INFO("Written graph to ", path);
+}
+
+std::vector<uint8_t> NetworkitBinaryWriter::writeToBuffer(const Graph &G) {
+    std::stringstream outdata;
+    writeData(outdata, G);
+    std::string str = outdata.str();
+    return std::vector<uint8_t>(str.begin(), str.end());
+}
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryWriter.cpp:35]: This incurs a further copy. If we use a writable memory mapped file, the copy could be avoided. It would be fine to keep this code for now (while changing it later to a writable memory mapped file). On the other hand, using a writable file and replacing `outStream.write` by `memcpy` does not seem to be that difficult either. Do you want to change that now or keep it as it is for this PR?

File: include/networkit/io/NetworkitBinaryGraph.hpp [hunk 10]
```diff
@@ -28,7 +28,9 @@ struct Header {
     uint64_t offsetAdjLists;
     uint64_t offsetAdjTranspose;
     uint64_t offsetWeightLists;
-    uint64_t offsetWeightTranspose;
+    uint64_t offsetWeightTranspose;
```
Comment (avdgrinten) [inline include/networkit/io/NetworkitBinaryGraph.hpp:31]: There is extra whitespace at the end of this file.

File: networkit/cpp/io/NetworkitBinaryGraph.md (hunk 3 above)
Comment (avdgrinten) [reply networkit/cpp/io/NetworkitBinaryGraph.md]: This still needs to be fixed.

File: networkit/test/test_graph.py [hunk 11]
```diff
@@ -253,5 +254,10 @@ def testAddEdgeWithMissingNodes(self):
 		self.assertEqual(G.numberOfNodes(), 3)
 		self.assertEqual(G.numberOfEdges(), 2)
 
+	def testGraphPickling(self):
+		G = nk.Graph(2)
+		G.addEdge(0,1)
+		pickle.dumps(G)
```
Comment (avdgrinten) [inline networkit/test/test_graph.py:260]: The test should also verify that unpickling works and that the deseralized graph is identical to the serialized one.

File: include/networkit/io/NetworkitBinaryWriter.hpp [hunk 12]
```diff
@@ -30,13 +30,19 @@ class NetworkitBinaryWriter final : public GraphWriter {
 
 public:
     NetworkitBinaryWriter(uint64_t chunks = 32,
-                          NetworkitBinaryWeights weightsType = NetworkitBinaryWeights::autoDetect);
+                          NetworkitBinaryWeights weightsType = NetworkitBinaryWeights::autoDetect,
+                          bool preserveEdgeIndex = false);
```
Comment (avdgrinten) [inline include/networkit/io/NetworkitBinaryWriter.hpp:34]: Why is this `false` by default? I think we should default to true if the graph has edge indices.

What happens if the users sets `true` here but the graph has no edge indices?

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 13]
```diff
@@ -106,26 +133,34 @@ Graph NetworkitBinaryReader::read(const std::string& path) {
...
+        uint64_t transIndexOff = 0;
         if(vertex) {
             memcpy(&off, adjIt + (c-1)* sizeof(uint64_t), sizeof(uint64_t));
             memcpy(&transpOff, transpIt + (c-1)* sizeof(uint64_t), sizeof(uint64_t));
             memcpy(&wghtOff, adjWghtIt + (c-1) * sizeof(uint64_t), sizeof(uint64_t));
             memcpy(&transWghtOff, transpWghtIt + (c-1) * sizeof(uint64_t), sizeof(uint64_t));
+            if(indexed){
+            memcpy(&indexOff, adjIdIt + (c-1) * sizeof(uint64_t), sizeof(uint64_t));
+            memcpy(&transIndexOff, transpIdIt + (c-1) * sizeof(uint64_t), sizeof(uint64_t));
+            }
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:153]: Doesn't match NetworKit's coding style.

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 14]
```diff
@@ -106,26 +133,34 @@ Graph NetworkitBinaryReader::read(const std::string& path) {
...
         transpOff += (chunks - 1) * sizeof(uint64_t);
         off += sizeof(uint64_t);
         transpOff += sizeof(uint64_t);
         wghtOff += (chunks - 1) * sizeof(uint64_t);
         transWghtOff += (chunks - 1) * sizeof(uint64_t);
+        indexOff += (chunks - 1) * sizeof(uint64_t);
+        transIndexOff += (chunks - 1) * sizeof(uint64_t);
         uint64_t n = firstVert[c+1] - firstVert[c];
-
         for (uint64_t i = 0; i < n; i++) {
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:164]: Unrelated change.

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 15]
```diff
@@ -216,12 +258,17 @@ Graph NetworkitBinaryReader::read(const std::string& path) {
...
                         break;
                 }
+                if(indexed){
+                    transIndexOff += nkbg::varIntDecode(reinterpret_cast<const uint8_t*>(transpIdIt + transIndexOff), id);
+                }
                 if(!directed) {
                     if(curr != add) {
-                        G.addPartialEdge(unsafe, curr, add, weight);
+                        if(indexed) G.addPartialEdge(unsafe, curr, add, weight, id);
+                        else G.addPartialEdge(unsafe, curr, add, weight);
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:267]: This calls the same function, right? We can get rid of the `if`.

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 16]
```diff
@@ -216,12 +258,17 @@ Graph NetworkitBinaryReader::read(const std::string& path) {
...
                     }
                 } else {
-                    G.addPartialInEdge(unsafe, curr, add, weight);
+                    if(indexed) G.addPartialInEdge(unsafe, curr, add, weight, id);
+                    else G.addPartialInEdge(unsafe, curr, add, weight);
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:271]: Likewise.

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 17]
```diff
@@ -232,6 +279,7 @@ Graph NetworkitBinaryReader::read(const std::string& path) {
     for(omp_index c = 0; c < static_cast<omp_index>(chunks); c++) {
         constructGraph(c);
     }
+
     G.setNumberOfSelfLoops(unsafe, selfLoops);
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:283]: Unrelated change.

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 18]
```diff
@@ -172,10 +208,15 @@ Graph NetworkitBinaryReader::read(const std::string& path) {
                     case nkbg::WEIGHT_FORMAT::NONE:
                         break;
                 }
+                if(indexed){
+                    indexOff += nkbg::varIntDecode(reinterpret_cast<const uint8_t*>(adjIdIt + indexOff), id);
+                }
                 if(!directed) {
-                    G.addPartialEdge(unsafe, curr, add, weight);
+                    if(indexed) G.addPartialEdge(unsafe, curr, add, weight, id);
+                    else G.addPartialEdge(unsafe, curr, add, weight);
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:216]: See below.

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 19]
```diff
@@ -172,10 +208,15 @@ Graph NetworkitBinaryReader::read(const std::string& path) {
...
                 } else {
-                    G.addPartialOutEdge(unsafe, curr, add, weight);
+                    if(indexed) G.addPartialOutEdge(unsafe, curr, add, weight, id);
+                    else G.addPartialOutEdge(unsafe, curr, add, weight);
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:219]: See below.

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 20]
```diff
@@ -139,6 +174,7 @@ Graph NetworkitBinaryReader::read(const std::string& path) {
             }
             //Read adjacency lists.
             for (uint64_t j = 0; j < outNbrs; j++) {
+                edgeid id;
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:177]: Don't we have to set the `Graph`'s maximal edge ID somewhere? Otherwise the `Graph` will allocate the same ID again when adding new edges.

File: networkit/cpp/io/NetworkitBinaryWriter.cpp [hunk 21]
```diff
@@ -180,6 +200,13 @@ void NetworkitBinaryWriter::write(const Graph &G, const std::string &path) {
         return size;
     };
 
+    auto computeIndexOffsets = [&] (edgeid id) {
+        uint64_t size = 0;
+        uint8_t tmp [10];
+        if(preserveEdgeIndex) size = nkbg::varIntEncode(id, tmp);
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryWriter.cpp:206]: Doesn't match NetworKit's coding style.

File: networkit/cpp/io/NetworkitBinaryWriter.cpp [hunk 22]
```diff
@@ -191,23 +218,27 @@ void NetworkitBinaryWriter::write(const Graph &G, const std::string &path) {
                         outNbrs++;
                         adjSize += nkbg::varIntEncode(v, tmp);
                         adjWeightSize += computeWeightsOffsets(w);
+                        if(preserveEdgeIndex) adjIndexSize += computeIndexOffsets(G.edgeId(n, v));
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryWriter.cpp:221]: Coding style, also below.

File: networkit/cpp/io/NetworkitBinaryWriter.cpp [hunk 23]
```diff
@@ -222,32 +253,50 @@ void NetworkitBinaryWriter::write(const Graph &G, const std::string &path) {
         transpOffsets.push_back(transpSize);
         adjWghtOffsets.push_back(adjWeightSize);
         transpWghtOffsets.push_back(transpWeightSize);
+        adjIndexOffsets.push_back(adjIndexSize);
+        transpIndexOffsets.push_back(transpIndexSize);
     }
     // Write header.
-    strncpy(header.magic,"nkbg002",8);
+    strncpy(header.magic,"nkbg002", 8);
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryWriter.cpp:260]: Unrelated change.

File: networkit/cpp/io/NetworkitBinaryWriter.cpp [hunk 24]
```diff
@@ -347,7 +397,51 @@ void NetworkitBinaryWriter::write(const Graph &G, const std::string &path) {
...
     });
-
-    INFO("Written graph to ", path);
+    
+    // Write adj index.
+    for (uint64_t c = 1; c < chunks; c++) {
+        outStream.write(reinterpret_cast<char*>(&adjIndexOffsets[c-1]), sizeof(uint64_t));
+    }
+    if(preserveEdgeIndex){
+    G.forNodes([&](node u) {
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryWriter.cpp:406]: Coding style.

**Comment by avdgrinten** [comment]
The commit structure is a bit messy. Squash all the changes to the C++ reader and writer to a single commit. The commits that deal with tests and/or the Python interface can stay separate.

File: networkit/cpp/io/NetworkitBinaryReader.cpp (hunk 20 above)
Comment (Relux-the-Relux) [reply networkit/cpp/io/NetworkitBinaryReader.cpp:177]: No, because we are reading the ids from file, not generating ourselves. See line 213.

File: include/networkit/io/NetworkitBinaryWriter.hpp [hunk 25]
```diff
@@ -30,13 +30,20 @@ class NetworkitBinaryWriter final : public GraphWriter {
 
 public:
     NetworkitBinaryWriter(uint64_t chunks = 32,
-                          NetworkitBinaryWeights weightsType = NetworkitBinaryWeights::autoDetect);
+                          NetworkitBinaryWeights weightsType = NetworkitBinaryWeights::autoDetect,
+                          bool ignoreEdgeIndex = false);
```
Comment (avdgrinten) [inline include/networkit/io/NetworkitBinaryWriter.hpp:34]: I don't like this `bool` flag here. Do we need it?

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 26]
```diff
@@ -45,6 +63,10 @@ Graph NetworkitBinaryReader::read(const std::string& path) {
         it += sizeof(uint64_t);
         memcpy(&header.offsetWeightTranspose, it, sizeof(uint64_t));
         it += sizeof(uint64_t);
+        memcpy(&header.offsetAdjIdLists, it, sizeof(uint64_t));
+        it += sizeof(uint64_t);
+        memcpy(&header.offsetAdjIdTranspose, it, sizeof(uint64_t));
+        it += sizeof(uint64_t);
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:69]: We need to make sure that these fields actually exist, do we? Old files will not have these entries.

Comment (avdgrinten) [reply networkit/cpp/io/NetworkitBinaryReader.cpp:69]: In fact, the header of old files will simply have the wrong size (since these fields are present whether edge IDs are in use or not). This probably requires a version bump.

Note that we still need to be able to read files of version 2 (and we should probably only write version 3 files if edge IDs are in use).

**Review by avdgrinten — COMMENTED** [review]
It would be nice to have this merged soon but there are a few remaining issue that I commented on below.

File: include/networkit/io/NetworkitBinaryWriter.hpp (hunk 25 above)
Comment (Relux-the-Relux) [reply include/networkit/io/NetworkitBinaryWriter.hpp:34]: Not really I could remove it and make always and only save indices of previously indexed graphs.

File: networkit/cpp/io/NetworkitBinaryReader.cpp (hunk 26 above)
Comment (Relux-the-Relux) [reply networkit/cpp/io/NetworkitBinaryReader.cpp:69]: I don't think it is necessary, since the size of the header shouldn't matter since we get a pointer(explicitly written) to where the data begins. It is the same case with unweighted graphs, where no weight data is written but we reader the header for it anyway, since it will be later handled with the header checker depending on the features flags.

yeah, so the size of the header doesn't matter because of the feature flags. since we will never read the pointers for IDs if that feature is not turned on, otherwise we just seek the data at the pointers specified in the header (not necessarily where the header immediately ends).

It is a similar situation to unweighted graphs where no weight data is written.

But I should test it.

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 27]
```diff
@@ -49,6 +49,21 @@ Graph NetworkitBinaryReader::readData(const T &source) {
...
+                weightFormat = static_cast<nkbg::WEIGHT_FORMAT>((header.features & nkbg::WGHT_MASK) >> nkbg::WGHT_SHIFT);
+                indexed = false;
+                version = 2;
+            }
+        } else {
+            directed = (header.features & nkbg::DIR_MASK);
+            weightFormat = static_cast<nkbg::WEIGHT_FORMAT>((header.features & nkbg::WGHT_MASK) >> nkbg::WGHT_SHIFT);
+            indexed = (header.features & nkbg::INDEX_MASK) >> nkbg::INDEX_SHIFT;
+            version = 3;
+        }
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:66]: There is some redundancy here. I think this part of the code should be restructured, for example like this:
```suggestion
if (!memcmp("nkbg002", header.magic, 8)) {
    version = 2;
}else if (!memcmp("nkbg003", header.magic, 8)) {
    version = 3;
}else{
    throw std::runtime_error("Reader expected another magic value");
}
directed = (header.features & nkbg::DIR_MASK);
weightFormat = static_cast<nkbg::WEIGHT_FORMAT>((header.features & nkbg::WGHT_MASK) >> nkbg::WGHT_SHIFT);
indexed = false;
if (version >= 3)
    indexed = (header.features & nkbg::INDEX_MASK) >> nkbg::INDEX_SHIFT;
```

File: networkit/cpp/io/NetworkitBinaryReader.cpp [hunk 28]
```diff
@@ -118,10 +125,8 @@ Graph NetworkitBinaryReader::readData(const T &source) {
     const char *transpIt = startIt + header.offsetAdjTranspose;
     const char *adjWghtIt = startIt + header.offsetWeightLists;
     const char *transpWghtIt = startIt + header.offsetWeightTranspose;
-    const char *adjIdIt = startIt + header.offsetAdjIdLists;
-    const char *transpIdIt = startIt + header.offsetAdjIdTranspose;
-
-
+    const char *adjIdIt = (version > 2) ? startIt + header.offsetAdjIdLists : 0;
+    const char *transpIdIt = (version > 2) ? startIt + header.offsetAdjIdTranspose : 0;
```
Comment (avdgrinten) [inline networkit/cpp/io/NetworkitBinaryReader.cpp:129]: Use `nullptr` here.

**Review by avdgrinten — COMMENTED** [review]
Can be merged after some minor fixes. When rebasing, please also change the commit messages to be in imperative: `expands Graph constructor ...` -> `Expand Graph constructor ...`

**Review by avdgrinten — APPROVED** [review]
LGTM.
